src/include/function/gds/gds_frontier.h

```diff
@@ -147,6 +140,67 @@ class GDSFrontier {
     common::table_id_map_t<common::offset_t> numNodesMap;
 };
 
+/*
+ * LocalFrontierSample keeps a small frontier per thread.
+ * If enabled, this frontier represents a complete local frontier.
+ * Otherwise, complete local frontier is larger than this sample.
+ * */
+class LocalFrontierSample {
+public:
+    bool enabled() const { return enabled_; }
+
+    const std::unordered_set<common::offset_t>& getOffsetSet() const { return offsetSet; }
+
+    bool empty() const { return offsetSet.empty(); }
+
+    void addNodes(const std::vector<common::nodeID_t> nodeIDs);
+
+    void checkSampleSize();
+
+private:
+    static constexpr uint16_t SAMPLE_SIZE = 20;
```

Comment:
This looks quite small to me. It also should probably depend on the number of nodes in the graph and threads. Can you implement a rule somewhere that is something like: SAMPLE_SIZE = max(20, |V|/(8xnumThreads)). To explain |V|/(8xnumThreads): The intuition is that sparse execution has the advantage that it does not loop over |V| many nodes, so that goes in the numerator. The 8 x numThreads models the disadvantage of sparse execution, which is due to two components: (i) that it runs single threaded instead of running in numThreads; and (ii) it does some extra operations, such as updating both the sparse and the dense frontier. So I'm putting a constant 8 to accommodate for the latter. That is until V/(8xnumThreads)-size we're better of taking both costs and running in a sparse manner.

This is another case that requires some benchmarking. To understand what SAMPLE_SIZE should be, we need to see the behavior of different heuristics to pick a SAMPLE_SIZE on some benchmark.

Reply:
I tried to make sample size proportional to maxOffset but this wasn't a trivial change. Also as we discussed, we might wanna experiment more on this. So instead I opened an issue.